File: shellheim/src/types/identity.ts
// Identity (credential) types

export interface Identity {
  id: string;
  account_id: string;
  name: string;
  username?: string;
  // Note: encrypted fields are not exposed to frontend
  // password_encrypted, ssh_key_encrypted, passphrase_encrypted are excluded
  created_at: string;
  updated_at: string;
}

export interface CreateIdentityRequest {
  name: string;
  username?: string;
  password?: string;
  ssh_key?: string;
  passphrase?: string;
}

export interface UpdateIdentityRequest {
  name?: string;
  username?: string;
  password?: string;
  ssh_key?: string;
  passphrase?: string;
}

// Credential type for UI display
export type CredentialType = 'password' | 'ssh_key' | 'both';

// Helper to determine what credentials an identity has
export function getCredentialType(identity: Identity & { hasPassword?: boolean; hasSshKey?: boolean }): CredentialType {
  const hasPassword = identity.hasPassword ?? false;
  const hasSshKey = identity.hasSshKey ?? false;
  
  if (hasPassword && hasSshKey) return 'both';
  if (hasSshKey) return 'ssh_key';
  return 'password';
}

File: shellheim/src-tauri/src/models/identity.rs
//! Identity (credentials) model

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Identity {
    pub id: String,
    pub account_id: String,
    pub name: String,
    pub username: Option<String>,
    #[serde(skip_serializing)]
    pub password_encrypted: Option<String>,
    #[serde(skip_serializing)]
    pub ssh_key_encrypted: Option<String>,
    #[serde(skip_serializing)]
    pub passphrase_encrypted: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Identity with decrypted fields (for internal use)
#[derive(Debug, Clone)]
pub struct DecryptedIdentity {
    pub id: String,
    pub name: String,
    pub username: Option<String>,
    pub password: Option<String>,
    pub ssh_key: Option<String>,
    pub passphrase: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateIdentityRequest {
    pub name: String,
    pub username: Option<String>,
    pub password: Option<String>,
    pub ssh_key: Option<String>,
    pub passphrase: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateIdentityRequest {
    pub name: Option<String>,
    pub username: Option<String>,
    pub password: Option<String>,
    pub ssh_key: Option<String>,
    pub passphrase: Option<String>,
}
